modulemd/modulemd-validator.1
.TH MODULEMD-VALIDATOR "1" "" "libmodulemd" "User Commands"
.SH NAME
modulemd-validator \- validate modulemd documents
.SH SYNOPSIS
.SY modulemd\-validator
[\fB\-\-quiet\fP|\fB\-\-verbose\fP|\fB\-\-debug\fP]
\fIFILE\fP\&.\|.\|.\&
.SY modulemd\-validator
[\fB\-\-quiet\fP|\fB\-\-verbose\fP|\fB\-\-debug\fP]
\fB\-\-type\fP=\fITYPE\fP
.I FILE
.SY modulemd\-validator
{\fB\-\-version\fP|\fB\-\-help\fP}
.YS
.SH DESCRIPTION
This tool validates modulemd documents specified with the positional arguments
using a libmodulemd parser.
.P
Without a \fB--type\fP option, all document types which can be found in a\~YUM
repository will be recognized. In this case each \fIFILE\fP can contain
multiple modulemd documents.
.P
Otherwise, exactly one \fIFILE\fP containing exactly one modulemd document must be
provided and the document will be validated against the given \fITYPE\fP.
.SH OPTIONS
.TP
\fB\-\-type\fP=\fI\,TYPE\/\fP
Constrain a document type. The recognized types are:
.RS
.TP
.B modulemd\-v1
An obsolete format used for building and installing modules.
.TP
.B modulemd\-v2
A deprecated format for building modules and a contemporary format for built
modules which can be placed into a\~YUM repository and installed by a package
manager.
.TP
.B modulemd\-defaults\-v1
A format for defining default streams and default profiles of a module. Found
in YUM repositories.
.TP
.B modulemd\-obsoletes\-v1
A format for defining streams which reach or will reach its end of
life. It can also recommend a new stream which obsoleted the old one. Found in
YUM repositories.
.TP
.B modulemd\-packager\-v2
An obsolete format for building modules.
.TP
.B modulemd\-packager\-v3
A contemporary format for building modules.
.TP
.B modulemd\-translations\-v1
A contemporary format for delivering localized human-oriented texts (e.\|g.\&
a description or a summary) associated with a module. Found in YUM
repositories.
.RE
.IP
Only single\-document files are supported.
For compatibility, if this options is not provided, any document type loadable
into a modulemd index (i.\|e.\& those intended for YUM repositories) is
acceptable.
.TP
\fB\-\-debug\fP
Output debugging messages.
.TP
\fB\-q\fP, \fB\-\-quiet\fP
Print no output.
.TP
\fB\-v\fP, \fB\-\-verbose\fP
Be verbose.
.TP
\fB\-h\fP, \fB\-\-help\fP
Print a usage text and then exit.
.TP
\fB\-V\fP, \fB\-\-version\fP
Print a version number and then exit.
.SH DEPRECATION NOTICE
Support for compressed \fIFILES\fP is deprecated.
.SH EXIT CODE
If the supplied documents are valid (and satisfy the \fB--type\fP constrain),
zero exit code will be returned.  Otherwise a non-zero code will be emitted.
.SH REPORTING BUGS
Bugs and feature requests can be reported at the
.UR https\://github.com/fedora-modularity/libmodulemd
libmodulemd home page
.UE .
.SH SEE ALSO
Specification of the modulemd documents is available in a \fByaml_spec\fP directory at the
libmodulemd home page.



NEWS
Release history for libmodulemd

2.15.1      2025-05-09

Fixes:
- A new "g_variant_store: assertion 'data != NULL' failed" warning added to
  GLib 2.84.1 when processing modulemd YAML documents with /data/xmd field with
  an empty flow map ({}) value was fixed.
- Building documentation now works with GLib 2.80.1 at the expense of missing
  cross links from libmodulemd to GLib. If it still fails you, you can disable
  building the documentation with "-Dwith_docs=false" meson option.
- A number of warnings printed by recent Meson was reduced.


2.15.0      2023-05-10

Enhancements:
- Add a support for decompressing Zstandard-compressed YAML files.
  modulemd_detect_compression() function now recognizes files with "zst" file
  name extension and files with Zstandard magic bytes and returns
  MODULEMD_COMPRESSION_TYPE_ZSTD_COMPRESSION constant for them.
  modulemd_module_index_update_from_file() can load Zstandard-compressed YAML
  files now.
- Remove a dependency on "file" library (libmagic). libmodulemd now implements
  its own detection of supported compression formats. This saves roughly 9 MB
  from a minimal system installation.
- An XML specification for in-YUM-repository metadata is available in
  "xml_specs" directory. This new format aspires for replacing modulemd-v2,
  modulemd-defaults-v1 and modulemd-translations-v1 YAML formats. A motivation
  is to unify with other XML files in YUM repositories. Please note that this
  specification is not yet implemented.

Incompatible changes:
- Meson minimal version increased to 0.55.

Deprecation:
- Deprecate intents in modulemd-defaults specification. System intents, which
  parameterize a default stream and default profiles of a module, were never
  implemented in the package managers (e.g. DNF). Users are discouraged from
  using this feature. Future libmodulmed versions might deprecate and later
  remove parts of an API pertaining the intents.

Fixes:
- Fix a crash when converting a modulemd-packager object with a default
  profile and without a module name or stream to a module index.
- Fix parsing empty profiles. Previously a profile with an empty rpms list was
  misparsed as a list with a single, empty package name.
- A specification for modulemd-v2 format was corrected to require a "content"
  license subtree only if the module build contains artifacts.
- Double-quote strings in scalar YAML values when they look like a number.
  Most of the fields in the YAML files are required to be interpreted as
  strings. Unaware third-party parsers could misinterpret number-like strings
  (e.g. 2.30) as floats (i.e. 2.3), leading to mismatching string
  representations. This is now prevented by quoting these values (i.e.
  "2.30").
- A build script now correctly asserts a requested Python version. Previously
  Python 3 /usr/bin/python program could be mistaken with Python 2
  interpreter.
- meson-0.64.0 warnings about "check" option of run_command() were fixed.
- Warnings from g_str_equal() macro of glib2 about passing an unsigned
  char * to strcmp() were fixed.


2.14.0      2022-02-04

This release is fully compatible with the previous version 2.13.0 except
behaviour changes described in Fixes section.

Enhancements:
- New functions for stripping XMD from an index
  (modulemd_module_index_clear_xmds()), from a module
  (modulemd_module_clear_xmds()), and from a version 2 module stream
  (modulemd_module_stream_v2_clear_xmd()). This enables you to prevent from leaking
  private build options to output modulemd-v2 documents.
- Interpret an invalid buildorder 18446744073709551615 as -1 if the library
  is built with a new build boolean accept_overflowed_buildorder set to true.
- Document module version, buildorder, and epoch limits in the specifications.
  Also clarify that an epoch number at RPM artifacts is mandatory.
- Document that modulemd_module_get_defaults can return NULL
- Improve a documentation for modulemd_module_set_defaults() and
  modulemd_module_index_upgrade_defaults() functions.
- Improve modulemd-validator(1) manual page.
- Drop a build dependency on help2man tool.

Fixes:
- Fix a crash when updating an index with an invalid subdocument and a NULL
  error parameter.
- Fix clobbering module defaults on a modulemd_module_set_defaults() error.
- Fix modulemd_module_index_upgrade_defaults() to actually use the requested
  version.
- Reject duplicate contexts in modulemd-packager-v3 documents.
- Report an error if modulemd-validator tool is invoked with both --version
  option and an unknown option.
- Fix an undefined behaviour when comparing integers of different signs.
- Fix DNF examples in a documentation.
- Actually execute modulemd-obsoletes Python tests.
- Fix deprecation warnings in tests with glib2-2.69.0.
- Fix deprecation warnings in Python tests and make the test compatible with
  Python 3.11.


2.13.0      2021-07-09

This release is fully compatible with the previous 2.12.1 version.

Enhancements:
- Add /data/demodularized/rpms list to modulemd and modulemd-packager formats.
  The listed binary RPM package names will be handled by a package manager as
  removed from all previous module versions. This enables moving a package from
  a module to a non-modular package set.
- modulemd-validator tool enables you to constrain a document type with a new
  "--type" option. Without this option, all document types recognized by
  a modulemd index are accepted. To validate version 3 of modulemd-packager
  format, use "modulemd-validator --type modulemd-packager-v3 foo.yaml"
  command.
- Allow disabling Python 3. You can pass -Dwith_py3=false to a meson setup
  command to disable building a binding for Python 3. The binding is enabled
  by default.

Fixes:
- Reject invalid integers. Purely non-numeric values and negative numbers where
  an unsigned type is mandated by a specification raise a parser error now.
- Handle a failed g_setenv() call in modulemd-validator.

modulemd/modulemd-validator.c
/*
 * This file is part of libmodulemd
 * Copyright 2018 Stephen Gallagher
 *
 * Fedora-License-Identifier: MIT
 * SPDX-2.0-License-Identifier: MIT
 * SPDX-3.0-License-Identifier: MIT
 *
 * This program is free software.
 * For more information on the license, see COPYING.
 * For more information on free software, see <https://www.gnu.org/philosophy/free-sw.en.html>.
 */


#include "modulemd.h"
#include "modulemd-errors.h"
#include "private/modulemd-defaults-v1-private.h"
#include "private/modulemd-module-index-private.h"
#include "private/modulemd-module-stream-v1-private.h"
#include "private/modulemd-module-stream-v2-private.h"
#include "private/modulemd-obsoletes-private.h"
#include "private/modulemd-subdocument-info-private.h"
#include "private/modulemd-translation-private.h"
#include "private/modulemd-yaml.h"
#include "private/modulemd-util.h"

#include <errno.h>
#include <glib.h>
#include <locale.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

enum mmd_verbosity
{
  MMD_QUIET = -1,
  MMD_DEFAULT,
  MMD_VERBOSE,
  MMD_DEBUG
};

/* Identifiers for modulemd type-version documents.
 * We cannot use GTypes, e.g. MODULEMD_TYPE_PACKAGER_V3, because
 * modulemd-packager-v2 GType does not exist (the format is parsed into
 * a modulemd-v2 data structure). Also GTypes are not constants. They are
 * a run-time computed value and hence cannot be used in switch-cases. */
enum mmd_type
{
  MMD_TYPE_INDEX, /* Untyped validation by loading into an index */
  MMD_TYPE_MODULEMD_V1,
  MMD_TYPE_MODULEMD_V2,
  MMD_TYPE_MODULEMD_DEFAULTS_V1,
  MMD_TYPE_MODULEMD_OBSOLETES_V1,
  MMD_TYPE_MODULEMD_PACKAGER_V2,
  MMD_TYPE_MODULEMD_PACKAGER_V3,
  MMD_TYPE_MODULEMD_TRANSLATIONS_V1
};

struct validator_options
{
  enum mmd_verbosity verbosity;
  GType type;
  gchar **filenames;
  gboolean report_version;
};

struct validator_options options = { 0 };

static gboolean
set_report_version (const gchar *UNUSED (option_name),
                    const gchar *UNUSED (value),
                    gpointer UNUSED (data),
                    GError **UNUSED (error))
{
  options.report_version = TRUE;
  return TRUE;
}

static gboolean
set_verbosity (const gchar *option_name,
               const gchar *UNUSED (value),
               gpointer UNUSED (data),
               GError **error)
{
  g_autofree gchar *debugging_env = NULL;
  if (g_strcmp0 ("-v", option_name) == 0 ||
      g_strcmp0 ("--verbose", option_name) == 0)
    {
      if (options.verbosity < MMD_VERBOSE)
        {
          options.verbosity = MMD_VERBOSE;
        }
    }
  else if (g_strcmp0 ("--debug", option_name) == 0)
    {
      if (options.verbosity < MMD_DEBUG)
        {
          options.verbosity = MMD_DEBUG;
          const gchar *old_debug = g_getenv ("G_MESSAGES_DEBUG");
          if (old_debug != NULL)
            {
              debugging_env =
                g_strdup_printf ("%s,%s", old_debug, G_LOG_DOMAIN);
            }
          else
            {
              debugging_env = g_strdup (G_LOG_DOMAIN);
            }
          if (!g_setenv ("G_MESSAGES_DEBUG", debugging_env, TRUE))
            {
              g_set_error (
                error,
                G_OPTION_ERROR,
                G_OPTION_ERROR_FAILED,
                "Could not set G_MESSAGES_DEBUG environment variable");
              return FALSE;
            }
        }
    }
  else if (g_strcmp0 ("-q", option_name) == 0 ||
           g_strcmp0 ("--quiet", option_name) == 0)
    {
      options.verbosity = MMD_QUIET;
    }
  else
    {
      /* We shouldn't be called under any other circumstance */
      g_set_error (error,
                   G_OPTION_ERROR,
                   G_OPTION_ERROR_FAILED,
                   "Called for unknown option \"%s\"",
                   option_name);
      return FALSE;
    }
  return TRUE;
}

static gboolean
set_type (const gchar *UNUSED (option_name),
          const gchar *value,
          gpointer UNUSED (data),
          GError **error)
{
  if (!g_strcmp0 (value, "modulemd-v1"))
    options.type = MMD_TYPE_MODULEMD_V1;
  else if (!g_strcmp0 (value, "modulemd-v2"))
    options.type = MMD_TYPE_MODULEMD_V2;
  else if (!g_strcmp0 (value, "modulemd-defaults-v1"))
    options.type = MMD_TYPE_MODULEMD_DEFAULTS_V1;
  else if (!g_strcmp0 (value, "modulemd-obsoletes-v1"))
    options.type = MMD_TYPE_MODULEMD_OBSOLETES_V1;
  else if (!g_strcmp0 (value, "modulemd-packager-v2"))
    options.type = MMD_TYPE_MODULEMD_PACKAGER_V2;
  else if (!g_strcmp0 (value, "modulemd-packager-v3"))
    options.type = MMD_TYPE_MODULEMD_PACKAGER_V3;
  else if (!g_strcmp0 (value, "modulemd-translations-v1"))
    options.type = MMD_TYPE_MODULEMD_TRANSLATIONS_V1;
  else
    {
      g_set_error (error,
                   G_OPTION_ERROR,
                   G_OPTION_ERROR_FAILED,
                   "Unknown document type: %s",
                   value);
      return FALSE;
    }
  return TRUE;
}

static const gchar *
mmd_type2astring (enum mmd_type type)
{
  switch (type)
    {
    case MMD_TYPE_INDEX: return "an index";
    case MMD_TYPE_MODULEMD_V1: return "a modulemd-v1";
    case MMD_TYPE_MODULEMD_V2: return "a modulemd-v2";
    case MMD_TYPE_MODULEMD_DEFAULTS_V1: return "a modulemd-defaults-v1";
    case MMD_TYPE_MODULEMD_OBSOLETES_V1: return "a modulemd-obsoletes-v1";
    case MMD_TYPE_MODULEMD_PACKAGER_V2: return "a modulemd-packager-v2";
    case MMD_TYPE_MODULEMD_PACKAGER_V3: return "a modulemd-packager-v3";
    case MMD_TYPE_MODULEMD_TRANSLATIONS_V1:
      return "a modulemd-translations-v1";
    }
  return "an unknown document type";
}

// clang-format off
static GOptionEntry entries[] = {
  { "debug", 0, G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_verbosity, "Output debugging messages", NULL },
  { "type", 0, G_OPTION_FLAG_NONE, G_OPTION_ARG_CALLBACK, set_type, "Constrain a document type (modulemd-v1, modulemd-v2, modulemd-defaults-v1, modulemd-obsoletes-v1, modulemd-packager-v2, modulemd-packager-v3, modulemd-translations-v1); by default any document type loadable into a modulemd index is acceptable; this option only supports single-document files", "TYPE" },
  { "quiet", 'q', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_verbosity, "Print no output", NULL },
  { "verbose", 'v', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_verbosity, "Be verbose", NULL },
  { "version", 'V', G_OPTION_FLAG_NO_ARG, G_OPTION_ARG_CALLBACK, set_report_version, "Print version number, then exit", NULL },
  { G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &options.filenames, "Files to be validated", NULL },
  { 0 } };
// clang-format on

static const gchar *
ModulemdYamlDocumentTypeEnum2string (ModulemdYamlDocumentTypeEnum type)
{
  switch (type)
    {
    case MODULEMD_YAML_DOC_MODULESTREAM: return "modulemd";
    case MODULEMD_YAML_DOC_DEFAULTS: return "modulemd-defaults";
    case MODULEMD_YAML_DOC_TRANSLATIONS: return "modulemd-translations";
    case MODULEMD_YAML_DOC_PACKAGER: return "modulemd-packager";
    case MODULEMD_YAML_DOC_OBSOLETES: return "modulemd-obsoletes";
    case MODULEMD_YAML_DOC_UNKNOWN: /* fall through */
    default: return "unknown type";
    }
}

/* We cannot load by index as it converts from old versions before
 * a return and as it does not provide enumeration functions for
 * subdocuments. We will use private modulemd_defaults_v1_parse_yaml() etc.
 * parsers. */
static gboolean
parse_file_as_subdoc_and_validate (const gchar *filename,
                                   enum mmd_type validation_type,
                                   ModulemdYamlDocumentTypeEnum expected_type,
                                   guint64 expected_version,
                                   GError **error)
{
  g_autoptr (FILE) file = NULL;
  MMD_INIT_YAML_PARSER (parser);
  MMD_INIT_YAML_EVENT (event);
  g_autoptr (ModulemdSubdocumentInfo) subdoc = NULL;
  const GError *subdoc_error;
  ModulemdYamlDocumentTypeEnum type;
  guint64 version;
  g_autoptr (GObject) object = NULL;

  /* Open the file and determine a document type */
  file = fopen (filename, "r");
  if (!file)
    {
      g_set_error (error,
                   MODULEMD_YAML_ERROR,
                   MMD_YAML_ERROR_OPEN,
                   "Could not open %s file: %s",
                   filename,
                   strerror (errno));
      return FALSE;
    }
  yaml_parser_set_input_file (&parser, file);
  if (!yaml_parser_parse (&parser, &event))
    {
      MMD_YAML_ERROR_EVENT_EXIT_BOOL (error, event, "Invalid YAML");
    }
  if (event.type != YAML_STREAM_START_EVENT)
    {
      {
        MMD_YAML_ERROR_EVENT_EXIT_BOOL (
          error, event, "YAML parser could not find a start of a YAML stream");
      }
    }
  yaml_event_delete (&event);
  if (!yaml_parser_parse (&parser, &event))
    {
      MMD_YAML_ERROR_EVENT_EXIT_BOOL (error, event, "Invalid YAML");
    }
  if (event.type != YAML_DOCUMENT_START_EVENT)
    {
      {
        MMD_YAML_ERROR_EVENT_EXIT_BOOL (
          error,
          event,
          "YAML parser could not find a start of a YAML document");
      }
    }
  yaml_event_delete (&event);
  subdoc = modulemd_yaml_parse_document_type (&parser);
  subdoc_error = modulemd_subdocument_info_get_gerror (subdoc);
  if (subdoc_error)
    {
      *error = g_error_copy (subdoc_error);
      return FALSE;
    }
  type = modulemd_subdocument_info_get_doctype (subdoc);
  if (type != expected_type)
    {
      g_set_error (error,
                   MODULEMD_ERROR,
                   0,
                   "Not %s document; it is %s",
                   mmd_type2astring (validation_type),
                   ModulemdYamlDocumentTypeEnum2string (type));
      return FALSE;
    }
  version = modulemd_subdocument_info_get_mdversion (subdoc);
  if (version != expected_version)
    {
      g_set_error (error,
                   MODULEMD_ERROR,
                   0,
                   "Not %s document; it is %" G_GUINT64_FORMAT " version",
                   mmd_type2astring (validation_type),
                   version);
      return FALSE;
    }

  /* Parse and validate the document body */
  switch (validation_type)
    {
    case MMD_TYPE_MODULEMD_V1:
      object =
        G_OBJECT (modulemd_module_stream_v1_parse_yaml (subdoc, TRUE, error));
      if (object && !modulemd_module_stream_validate (
                      MODULEMD_MODULE_STREAM (object), error))
        g_clear_object (&object);
      break;
    case MMD_TYPE_MODULEMD_DEFAULTS_V1:
      object =
        G_OBJECT (modulemd_defaults_v1_parse_yaml (subdoc, TRUE, error));
      /* validated implicitly */
      break;
    case MMD_TYPE_MODULEMD_OBSOLETES_V1:
      object = G_OBJECT (modulemd_obsoletes_parse_yaml (subdoc, TRUE, error));
      /* validated implicitly */
      break;
    case MMD_TYPE_MODULEMD_PACKAGER_V2:
      object = G_OBJECT (
        modulemd_module_stream_v2_parse_yaml (subdoc, TRUE, TRUE, error));
      if (object && !modulemd_module_stream_validate (
                      MODULEMD_MODULE_STREAM (object), error))
        g_clear_object (&object);
      break;
    case MMD_TYPE_MODULEMD_TRANSLATIONS_V1:
      object =
        G_OBJECT (modulemd_translation_parse_yaml (subdoc, TRUE, error));
      /* validated implicitly */
      break;
    default:
      g_set_error (error,
                   MODULEMD_ERROR,
                   0,
                   "Internal error: %s type is not supported",
                   mmd_type2astring (validation_type));
      return FALSE;
    }
  if (!object)
    return FALSE;

  /* Check for a garbage past the first document */
  if (!yaml_parser_parse (&parser, &event))
    {
      MMD_YAML_ERROR_EVENT_EXIT_BOOL (
        error, event, "Invalid YAML after first document");
    }
  if (event.type != YAML_STREAM_END_EVENT)
    {
      {
        MMD_YAML_ERROR_EVENT_EXIT_BOOL (
          error, event, "Another YAML document after the first one");
        return FALSE;
      }
    }
  yaml_event_delete (&event);
  return TRUE;
}

static gboolean
parse_file (const gchar *filename, GPtrArray **failures, GError **error)
{
  if (options.verbosity >= MMD_VERBOSE)
    {
      g_fprintf (stdout, "Validating %s\n", filename);
    }

  switch (options.type)
    {
    case MMD_TYPE_INDEX:
      {
        g_autoptr (ModulemdModuleIndex) index = NULL;
        index = modulemd_module_index_new ();
        return modulemd_module_index_update_from_file_ext (
          index, filename, TRUE, TRUE, failures, error);
      }
    case MMD_TYPE_MODULEMD_V1:
      return parse_file_as_subdoc_and_validate (
        filename, options.type, MODULEMD_YAML_DOC_MODULESTREAM, 1u, error);
    case MMD_TYPE_MODULEMD_V2:
      {
        GType type;
        g_autoptr (GObject) object = NULL;
        type = modulemd_read_packager_file (filename, &object, error);
        if (type == G_TYPE_INVALID)
          return FALSE;
        if (type != MODULEMD_TYPE_MODULE_STREAM_V2)
          {
            g_set_error (error,
                         MODULEMD_ERROR,
                         0,
                         "Not a modulemd-v2 document; it is %s",
                         g_type_name (type));
            return FALSE;
          }
        return modulemd_module_stream_validate (
          MODULEMD_MODULE_STREAM (object), error);
      }
    case MMD_TYPE_MODULEMD_DEFAULTS_V1:
      return parse_file_as_subdoc_and_validate (
        filename, options.type, MODULEMD_YAML_DOC_DEFAULTS, 1u, error);
    case MMD_TYPE_MODULEMD_OBSOLETES_V1:
      return parse_file_as_subdoc_and_validate (
        filename, options.type, MODULEMD_YAML_DOC_OBSOLETES, 1u, error);
    case MMD_TYPE_MODULEMD_PACKAGER_V2:
      return parse_file_as_subdoc_and_validate (
        filename, options.type, MODULEMD_YAML_DOC_PACKAGER, 2u, error);
    case MMD_TYPE_MODULEMD_PACKAGER_V3:
      {
        GType type;
        g_autoptr (GObject) object = NULL;
        type = modulemd_read_packager_file (filename, &object, error);
        if (type == G_TYPE_INVALID)
          return FALSE;
        if (type != MODULEMD_TYPE_PACKAGER_V3)
          {
            g_set_error (error,
                         MODULEMD_ERROR,
                         0,
                         "Not a modulemd-packager-v3 document; it is %s",
                         g_type_name (type));
            return FALSE;
          }
        /* modulemd_packager_v3 is validated implicitly by
         * modulemd_read_packager_file (). */
        return TRUE;
      }
    case MMD_TYPE_MODULEMD_TRANSLATIONS_V1:
      return parse_file_as_subdoc_and_validate (
        filename, options.type, MODULEMD_YAML_DOC_TRANSLATIONS, 1u, error);
    }
  g_fprintf (stderr,
             "Internal error: unsupported document type: %s\n",
             mmd_type2astring (options.type));
  exit (EXIT_FAILURE);
}


int
main (int argc, char *argv[])
{
  const char *filename;
  g_autoptr (GOptionContext) context = NULL;
  g_autoptr (GError) error = NULL;
  gsize num_invalid = 0;
  gboolean ret;
  g_autoptr (GPtrArray) failures = NULL;
  ModulemdSubdocumentInfo *doc = NULL;

  setlocale (LC_ALL, "");

  options.type = MMD_TYPE_INDEX;
  context = g_option_context_new ("FILES - Simple modulemd YAML validator");
  g_option_context_add_main_entries (context, entries, "modulemd-validator");
  if (!g_option_context_parse (context, &argc, &argv, &error))
    {
      g_fprintf (stderr, "option parsing failed: %s\n", error->message);
      exit (EXIT_FAILURE);
    }

  if (options.report_version)
    {
      g_fprintf (stdout, "modulemd-validator %s\n", modulemd_get_version ());
      exit (EXIT_SUCCESS);
    }

  if (!(options.filenames && options.filenames[0]))
    {
      g_fprintf (stderr,
                 "At least one file must be specified on the command-line\n");
      exit (EXIT_FAILURE);
    }

  for (gsize i = 0; options.filenames[i]; i++)
    {
      filename = options.filenames[i];

      ret = parse_file (filename, &failures, &error);
      if (!ret)
        {
          num_invalid++;
          if (options.verbosity >= MMD_DEFAULT)
            {
              g_fprintf (stderr, "%s failed to validate\n", filename);

              if (error != NULL)
                {
                  /* Unparseable content */
                  g_fprintf (stderr,
                             "%s could not be read in its entirety: %s\n",
                             filename,
                             error->message);
                }
              if (failures)
                {
                  for (gsize j = 0; j < failures->len; j++)
                    {
                      doc = MODULEMD_SUBDOCUMENT_INFO (
                        g_ptr_array_index (failures, j));
                      g_printf (
                        "\nFailed subdocument (%s): \n%s\n",
                        modulemd_subdocument_info_get_gerror (doc)->message,
                        modulemd_subdocument_info_get_yaml (doc));
                    }
                }
            }
        }
      else
        {
          if (options.verbosity >= MMD_DEFAULT)
            {
              g_printf ("%s validated successfully\n", filename);
            }
        }
      g_clear_error (&error);
      g_clear_pointer (&failures, g_ptr_array_unref);
    }

  return num_invalid;
}
